components/proxy/src/main/java/com/hotels/styx/proxy/BackendServicesRouter.java

```diff
@@ -154,20 +145,49 @@ public void onChange(Registry.Changes<BackendService> changes) {
                     .hostClientFactory(hostClientFactory)
                     .build();
 
-            pipeline = new ProxyToClientPipeline(newClientHandler(backendService, inventory, originStatsFactory), inventory);
+            pipeline = new ProxyToClientPipeline(newClientHandler(backendService, inventory, originStatsFactory), () -> {
+                inventory.close();
```

Comment:
It seems here (and above in `return runAsync(() -> {`... ) we might be not following the principle that says we should keep  statements in a function at the same level of abstraction.  Following up on that, it also seems the pipeline could contain the logic to close itself instead of transferring the responsibility to the caller.

Reply:
The idea was to avoid bloating the `ProxyToClientPipeline` constructor signature by providing a kind of listener that can react to `close` events, but I am open to alternatives.